src/runtime/formal/ops/ops.cc

```diff
@@ -42,27 +39,28 @@ CVM_REGISTER_GLOBAL("cvm.runtime.formal.dense")
   auto X_data = CVMArg2Data<int32_t>(X); 
   auto W_data = CVMArg2Data<int32_t>(W); 
   auto Y_data = CVMArg2Data<int32_t>(Y); 
-  // Y = X * WT
+  // Y = X * W^T
+  Indices xIdx(X_shape), wIdx(W_shape), yIdx(Y_shape);
   for (int64_t m = 0; m < Y_shape[0]; ++m) {
     // Y(m, n) = X(m, k) * WT(k, n) = X(m, k) * W(n, k)
-    int32_t Y_offset = m * Y_shape[1], X_offset = m * X_shape[1];
+    xIdx.Ref(0) = m;
     for (int64_t n = 0; n < Y_shape[1]; ++n) {
-      int32_t sum = 0, W_offset = n * W_shape[1];
+      wIdx.Ref(0) = n;
+      int32_t sum = 0;  //, W_offset = n * W_shape[1];
       for (int64_t k = 0; k < X_shape[1]; ++k) {
-        sum += X_data[X_offset + k] * W_data[W_offset + k];
+        xIdx.Ref(1) = k, wIdx.Ref(1) = k;
```

Comment:
You can simplify the `xIdx` and `wIdx` assignment with:

```suggestion
        xIdx.CopyIndicesFrom({m, k});
        wIdx.CopyIndicesFrom({n, k});
```

This index assignment is more intuitive and easier to understand.